rust/worker/src/segment/metadata_segment.rs

```diff
@@ -418,7 +481,210 @@ impl<'a> SegmentWriter<'a> for MetadataSegmentWriter {
+                                            match &self.string_metadata_index_writer {
+                                                Some(writer) => {
+                                                    let _ = writer.delete(
+                                                        key,
+                                                        value.as_str(),
+                                                        segment_offset_id,
+                                                    );
+                                                }
+                                                None => {}
+                                            }
+                                        }
+                                        MetadataValue::Float(value) => {
+                                            match &self.f32_metadata_index_writer {
+                                                Some(writer) => {
+                                                    let _ = writer.delete(
+                                                        key,
+                                                        *value as f32,
+                                                        segment_offset_id,
+                                                    );
+                                                }
+                                                None => {}
+                                            }
+                                        }
+                                        MetadataValue::Int(value) => {
+                                            match &self.u32_metadata_index_writer {
+                                                Some(writer) => {
+                                                    let _ = writer.delete(
+                                                        key,
+                                                        *value as u32,
+                                                        segment_offset_id,
+                                                    );
+                                                }
+                                                None => {}
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                            None => {}
+                        };
+                        match &data_record.document {
+                            Some(document) => match &self.full_text_index_writer {
+                                Some(writer) => {
+                                    let _ = writer.delete_document(document, segment_offset_id);
+                                }
+                                None => {}
+                            },
+                            None => {}
+                        };
                     }
+                    None => {}
+                },
+                Operation::Update => {
```

Comment:
Is this logic not handling deletes for a metadata key? Also, what about new metadata keys that were freshly inserted?

Reply:
For deletes we can add a TODO since I have not make the `None` values for the keys available in the materialized log records (which I haven't yet). For new keys I think we have all the support and it can be done now

Reply:
Discussed offline: Metadata may be added, updated, or deleted in an `Operation::Delete`. We need to check metadata keys in `metadata_to_be_merged` against `data_record.0.metadata`